File: sdk-typescript/packages/clips/src/clips.ts
/**
 * WAVE SDK - Clips API
 *
 * Create, manage, and export video clips from streams and recordings.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
  Timestamps,
  Metadata,
} from '@wave-av/core';
import type { ClipStatus, ClipExportFormat, ClipQualityPreset, ClipSource, Clip } from './clips-types';
export type { ClipStatus, ClipExportFormat, ClipQualityPreset, ClipSource, Clip } from './clips-types';

// ============================================================================
// Types
// ============================================================================

/**
 * Clip status
 */

/**
 * Clip export format
 */

/**
 * Clip quality preset
 */

/**
 * Clip source reference
 */

/**
 * Clip object
 */

/**
 * Create clip request
 */
export interface CreateClipRequest {
  title: string;
  description?: string;
  source: ClipSource;
  quality?: ClipQualityPreset;
  format?: ClipExportFormat;
  tags?: string[];
  metadata?: Metadata;
  /** Enable AI-powered highlight detection */
  auto_highlights?: boolean;
  /** Webhook URL for status updates */
  webhook_url?: string;
}

/**
 * Update clip request
 */
export interface UpdateClipRequest {
  title?: string;
  description?: string;
  tags?: string[];
  metadata?: Metadata;
}

/**
 * List clips filters
 */
export interface ListClipsParams extends PaginationParams {
  status?: ClipStatus;
  source_type?: 'stream' | 'recording' | 'upload';
  source_id?: string;
  tags?: string[];
  created_after?: string;
  created_before?: string;
  order_by?: 'created_at' | 'duration' | 'title';
  order?: 'asc' | 'desc';
}

/**
 * Export clip request
 */
export interface ExportClipRequest {
  format: ClipExportFormat;
  quality?: ClipQualityPreset;
  /** Custom resolution (e.g., "1920x1080") */
  resolution?: string;
  /** Custom bitrate in kbps */
  bitrate?: number;
  /** Include audio */
  include_audio?: boolean;
  /** Add watermark */
  watermark?: {
    image_url: string;
    position: 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right' | 'center';
    opacity?: number;
    scale?: number;
  };
}

/**
 * Export job status
 */
export interface ClipExport extends Timestamps {
  id: string;
  clip_id: string;
  status: 'pending' | 'processing' | 'ready' | 'failed';
  format: ClipExportFormat;
  download_url?: string;
  file_size?: number;
  expires_at?: string;
  error?: string;
}

/**
 * Auto-highlight result
 */
export interface ClipHighlight {
  start_time: number;
  end_time: number;
  score: number;
  type: 'action' | 'speech' | 'emotion' | 'custom';
  label?: string;
}

// ============================================================================
// Clips API
// ============================================================================

/**
 * Clips API client
 *
 * All operations require appropriate permissions. Authorization is enforced
 * server-side - the API returns 403 if the authenticated user lacks access.
 *
 * @example
 * ```typescript
 * import { WaveClient } from '@wave-av/core';
 * import { ClipsAPI } from '@wave-av/clips';
 *
 * const client = new WaveClient({ apiKey: 'your-api-key' });
 * const clips = new ClipsAPI(client);
 *
 * // Create a clip from a stream
 * const clip = await clips.create({
 *   title: 'Best Moment',
 *   source: {
 *     type: 'stream',
 *     id: 'stream_123',
 *     start_time: 120,
 *     end_time: 150,
 *   },
 * });
 *
 * // Wait for processing
 * const ready = await clips.waitForReady(clip.id);
 * console.log('Clip ready:', ready.playback_url);
 * ```
 */
export class ClipsAPI {
  private readonly client: WaveClient;
  private readonly basePath = '/v1/clips';

  constructor(client: WaveClient) {
    this.client = client;
  }

  /**
   * Create a new clip
   *
   * Requires: clips:create permission
   */
  async create(request: CreateClipRequest): Promise<Clip> {
    return this.client.post<Clip>(this.basePath, request);
  }

  /**
   * Get a clip by ID
   *
   * Requires: clips:read permission
   */
  async get(clipId: string): Promise<Clip> {
    return this.client.get<Clip>(`${this.basePath}/${clipId}`);
  }

  /**
   * Update a clip
   *
   * Requires: clips:update permission
   */
  async update(clipId: string, request: UpdateClipRequest): Promise<Clip> {
    return this.client.patch<Clip>(`${this.basePath}/${clipId}`, request);
  }

  /**
   * Remove a clip
   *
   * Requires: clips:remove permission (server-side RBAC enforced)
   */
  async remove(clipId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${clipId}`);
  }

  /**
   * List clips with optional filters
   *
   * Requires: clips:read permission
   */
  async list(params?: ListClipsParams): Promise<PaginatedResponse<Clip>> {
    const queryParams: Record<string, string | number | boolean | undefined> = {
      limit: params?.limit,
      offset: params?.offset,
      cursor: params?.cursor,
      status: params?.status,
      source_type: params?.source_type,
      source_id: params?.source_id,
      created_after: params?.created_after,
      created_before: params?.created_before,
      order_by: params?.order_by,
      order: params?.order,
    };

    if (params?.tags?.length) {
      queryParams['tags'] = params.tags.join(',');
    }

    return this.client.get<PaginatedResponse<Clip>>(this.basePath, {
      params: queryParams,
    });
  }

  /**
   * Export a clip to a different format
   *
   * Requires: clips:export permission
   */
  async exportClip(clipId: string, request: ExportClipRequest): Promise<ClipExport> {
    return this.client.post<ClipExport>(
      `${this.basePath}/${clipId}/export`,
      request
    );
  }

  /**
   * Get export job status
   *
   * Requires: clips:read permission
   */
  async getExport(clipId: string, exportId: string): Promise<ClipExport> {
    return this.client.get<ClipExport>(
      `${this.basePath}/${clipId}/exports/${exportId}`
    );
  }

  /**
   * List all exports for a clip
   *
   * Requires: clips:read permission
   */
  async listExports(
    clipId: string,
    params?: PaginationParams
  ): Promise<PaginatedResponse<ClipExport>> {
    return this.client.get<PaginatedResponse<ClipExport>>(
      `${this.basePath}/${clipId}/exports`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Detect highlights in source content
   *
   * Requires: clips:analyze permission
   */
  async detectHighlights(
    sourceType: 'stream' | 'recording',
    sourceId: string,
    options?: {
      types?: ('action' | 'speech' | 'emotion')[];
      min_score?: number;
      max_results?: number;
    }
  ): Promise<ClipHighlight[]> {
    return this.client.post<ClipHighlight[]>(
      `${this.basePath}/highlights/detect`,
      {
        source_type: sourceType,
        source_id: sourceId,
        ...options,
      }
    );
  }

  /**
   * Generate clips from detected highlights
   *
   * Requires: clips:create permission
   */
  async createFromHighlights(
    sourceType: 'stream' | 'recording',
    sourceId: string,
    options?: {
      min_score?: number;
      max_clips?: number;
      title_prefix?: string;
      tags?: string[];
    }
  ): Promise<Clip[]> {
    return this.client.post<Clip[]>(`${this.basePath}/highlights/create`, {
      source_type: sourceType,
      source_id: sourceId,
      ...options,
    });
  }

  /**
   * Wait for a clip to be ready
   */
  async waitForReady(
    clipId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (clip: Clip) => void;
    }
  ): Promise<Clip> {
    const pollInterval = options?.pollInterval || 2000;
    const timeout = options?.timeout || 300000; // 5 minutes default
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const clip = await this.get(clipId);

      if (options?.onProgress) {
        options.onProgress(clip);
      }

      if (clip.status === 'ready') {
        return clip;
      }

      if (clip.status === 'failed') {
        throw new Error(`Clip processing failed: ${clip.error || 'Unknown error'}`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Clip processing timed out after ${timeout}ms`);
  }

  /**
   * Wait for an export to be ready
   */
  async waitForExport(
    clipId: string,
    exportId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
    }
  ): Promise<ClipExport> {
    const pollInterval = options?.pollInterval || 2000;
    const timeout = options?.timeout || 300000;
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const exportJob = await this.getExport(clipId, exportId);

      if (exportJob.status === 'ready') {
        return exportJob;
      }

      if (exportJob.status === 'failed') {
        throw new Error(`Export failed: ${exportJob.error || 'Unknown error'}`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Export timed out after ${timeout}ms`);
  }
}

/**
 * Create a Clips API instance
 */
export function createClipsAPI(client: WaveClient): ClipsAPI {
  return new ClipsAPI(client);
}

File: sdk-python/wave_av_sdk/clips.py
"""
WAVE SDK - Clips API

Create, manage, and export video clips from streams and recordings.
"""

from __future__ import annotations

import time
from typing import Any, Literal
from pydantic import BaseModel

from wave.client import WaveClient


class ClipSource(BaseModel):
    """Clip source reference."""

    type: Literal["stream", "recording", "upload"]
    id: str
    start_time: float
    end_time: float


class Clip(BaseModel):
    """Clip object."""

    id: str
    organization_id: str
    title: str
    description: str | None = None
    source: ClipSource
    status: Literal["pending", "processing", "ready", "failed", "deleted"]
    duration: float
    thumbnail_url: str | None = None
    playback_url: str | None = None
    download_url: str | None = None
    file_size: int | None = None
    width: int | None = None
    height: int | None = None
    frame_rate: float | None = None
    bitrate: int | None = None
    codec: str | None = None
    tags: list[str] | None = None
    metadata: dict[str, Any] | None = None
    error: str | None = None
    created_at: str
    updated_at: str


class ClipExport(BaseModel):
    """Export job status."""

    id: str
    clip_id: str
    status: Literal["pending", "processing", "ready", "failed"]
    format: str
    download_url: str | None = None
    file_size: int | None = None
    expires_at: str | None = None
    error: str | None = None
    created_at: str
    updated_at: str


class ClipHighlight(BaseModel):
    """Auto-highlight result."""

    start_time: float
    end_time: float
    score: float
    type: Literal["action", "speech", "emotion", "custom"]
    label: str | None = None


class ClipsAPI:
    """
    Clips API client.

    Example:
        >>> from wave import Wave
        >>> wave = Wave(api_key="your-api-key")
        >>> clip = wave.clips.create(
        ...     title="Best Moment",
        ...     source={"type": "stream", "id": "stream_123", "start_time": 120, "end_time": 150}
        ... )
        >>> ready = wave.clips.wait_for_ready(clip.id)
    """

    def __init__(self, client: WaveClient):
        self._client = client
        self._base_path = "/v1/clips"

    def create(
        self,
        title: str,
        source: dict[str, Any] | ClipSource,
        description: str | None = None,
        quality: str | None = None,
        format: str | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        auto_highlights: bool | None = None,
        webhook_url: str | None = None,
    ) -> Clip:
        """
        Create a new clip.

        Args:
            title: Clip title
            source: Source reference with type, id, start_time, end_time
            description: Optional description
            quality: Quality preset (low, medium, high, source)
            format: Export format (mp4, webm, mov, gif)
            tags: Optional tags
            metadata: Custom metadata
            auto_highlights: Enable AI highlight detection
            webhook_url: Webhook URL for status updates

        Returns:
            Created clip object
        """
        data: dict[str, Any] = {"title": title, "source": source}
        if description is not None:
            data["description"] = description
        if quality is not None:
            data["quality"] = quality
        if format is not None:
            data["format"] = format
        if tags is not None:
            data["tags"] = tags
        if metadata is not None:
            data["metadata"] = metadata
        if auto_highlights is not None:
            data["auto_highlights"] = auto_highlights
        if webhook_url is not None:
            data["webhook_url"] = webhook_url

        response = self._client.post(self._base_path, json=data)
        return Clip(**response)

    def get(self, clip_id: str) -> Clip:
        """Get a clip by ID."""
        response = self._client.get(f"{self._base_path}/{clip_id}")
        return Clip(**response)

    def update(
        self,
        clip_id: str,
        title: str | None = None,
        description: str | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Clip:
        """Update a clip."""
        data: dict[str, Any] = {}
        if title is not None:
            data["title"] = title
        if description is not None:
            data["description"] = description
        if tags is not None:
            data["tags"] = tags
        if metadata is not None:
            data["metadata"] = metadata

        response = self._client.patch(f"{self._base_path}/{clip_id}", json=data)
        return Clip(**response)

    def remove(self, clip_id: str) -> None:
        """Remove a clip."""
        self._client.delete(f"{self._base_path}/{clip_id}")

    def list(
        self,
        status: str | None = None,
        source_type: str | None = None,
        source_id: str | None = None,
        tags: list[str] | None = None,
        created_after: str | None = None,
        created_before: str | None = None,
        order_by: str | None = None,
        order: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> dict[str, Any]:
        """List clips with optional filters."""
        params: dict[str, Any] = {}
        if status is not None:
            params["status"] = status
        if source_type is not None:
            params["source_type"] = source_type
        if source_id is not None:
            params["source_id"] = source_id
        if tags is not None:
            params["tags"] = ",".join(tags)
        if created_after is not None:
            params["created_after"] = created_after
        if created_before is not None:
            params["created_before"] = created_before
        if order_by is not None:
            params["order_by"] = order_by
        if order is not None:
            params["order"] = order
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        return self._client.get(self._base_path, params=params)

    def export_clip(
        self,
        clip_id: str,
        format: str,
        quality: str | None = None,
        resolution: str | None = None,
        bitrate: int | None = None,
        include_audio: bool | None = None,
        watermark: dict[str, Any] | None = None,
    ) -> ClipExport:
        """Export a clip to a different format."""
        data: dict[str, Any] = {"format": format}
        if quality is not None:
            data["quality"] = quality
        if resolution is not None:
            data["resolution"] = resolution
        if bitrate is not None:
            data["bitrate"] = bitrate
        if include_audio is not None:
            data["include_audio"] = include_audio
        if watermark is not None:
            data["watermark"] = watermark

        response = self._client.post(f"{self._base_path}/{clip_id}/export", json=data)
        return ClipExport(**response)

    def get_export(self, clip_id: str, export_id: str) -> ClipExport:
        """Get export job status."""
        response = self._client.get(f"{self._base_path}/{clip_id}/exports/{export_id}")
        return ClipExport(**response)

    def list_exports(
        self,
        clip_id: str,
        limit: int | None = None,
        offset: int | None = None,
    ) -> dict[str, Any]:
        """List all exports for a clip."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        return self._client.get(f"{self._base_path}/{clip_id}/exports", params=params)

    def detect_highlights(
        self,
        source_type: Literal["stream", "recording"],
        source_id: str,
        types: list[str] | None = None,
        min_score: float | None = None,
        max_results: int | None = None,
    ) -> list[ClipHighlight]:
        """Detect highlights in source content."""
        data: dict[str, Any] = {
            "source_type": source_type,
            "source_id": source_id,
        }
        if types is not None:
            data["types"] = types
        if min_score is not None:
            data["min_score"] = min_score
        if max_results is not None:
            data["max_results"] = max_results

        response = self._client.post(f"{self._base_path}/highlights/detect", json=data)
        return [ClipHighlight(**h) for h in response]

    def create_from_highlights(
        self,
        source_type: Literal["stream", "recording"],
        source_id: str,
        min_score: float | None = None,
        max_clips: int | None = None,
        title_prefix: str | None = None,
        tags: list[str] | None = None,
    ) -> list[Clip]:
        """Generate clips from detected highlights."""
        data: dict[str, Any] = {
            "source_type": source_type,
            "source_id": source_id,
        }
        if min_score is not None:
            data["min_score"] = min_score
        if max_clips is not None:
            data["max_clips"] = max_clips
        if title_prefix is not None:
            data["title_prefix"] = title_prefix
        if tags is not None:
            data["tags"] = tags

        response = self._client.post(f"{self._base_path}/highlights/create", json=data)
        return [Clip(**c) for c in response]

    def wait_for_ready(
        self,
        clip_id: str,
        poll_interval: float = 2.0,
        timeout: float = 300.0,
        on_progress: callable | None = None,
    ) -> Clip:
        """
        Wait for a clip to be ready.

        Args:
            clip_id: Clip ID to wait for
            poll_interval: Polling interval in seconds
            timeout: Maximum wait time in seconds
            on_progress: Optional callback for progress updates

        Returns:
            Ready clip object

        Raises:
            TimeoutError: If clip doesn't complete within timeout
            RuntimeError: If clip processing fails
        """
        start_time = time.time()

        while time.time() - start_time < timeout:
            clip = self.get(clip_id)

            if on_progress:
                on_progress(clip)

            if clip.status == "ready":
                return clip

            if clip.status == "failed":
                raise RuntimeError(f"Clip processing failed: {clip.error or 'Unknown error'}")

            time.sleep(poll_interval)

        raise TimeoutError(f"Clip processing timed out after {timeout}s")

    def wait_for_export(
        self,
        clip_id: str,
        export_id: str,
        poll_interval: float = 2.0,
        timeout: float = 300.0,
    ) -> ClipExport:
        """Wait for an export to be ready."""
        start_time = time.time()

        while time.time() - start_time < timeout:
            export = self.get_export(clip_id, export_id)

            if export.status == "ready":
                return export

            if export.status == "failed":
                raise RuntimeError(f"Export failed: {export.error or 'Unknown error'}")

            time.sleep(poll_interval)

        raise TimeoutError(f"Export timed out after {timeout}s")
